BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeError.kt

```diff
@@ -0,0 +1,179 @@
+package com.braintreepayments.api
+
+import android.os.Parcel
+import android.os.Parcelable
+import com.braintreepayments.api.GraphQLConstants.ErrorTypes
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+
+/**
+ * @property field Field name this object represents.
+ * @property message Human readable summary of the error for field. May be `null`.
+ * @property code Error code if one exists; defaults to [.UNKNOWN_CODE] otherwise
+ * @property fieldErrors [BraintreeError] objects for any errors nested under this field.
+ */
+open class BraintreeError : Parcelable {
+
+    open var field: String? = null
+        internal set
+
+    open var message: String? = null
+        internal set
+
+    open var fieldErrors: MutableList<BraintreeError>? = null
+        internal set
+
+    // default value
+    open var code = UNKNOWN_CODE
+        internal set
+
+    /**
+     * Method to extract an error for an individual field, e.g. creditCard, customer, etc.
+     *
+     * @param field Name of the field desired, expected to be in camelCase.
+     * @return [BraintreeError] for the field searched, or `null` if not found.
+     */
+    open fun errorFor(field: String): BraintreeError? {
+        var returnError: BraintreeError?
+        if (fieldErrors != null) {
+            for (error in fieldErrors!!) {
+                if (error.field == field) {
+                    return error
+                } else if (error.fieldErrors != null) {
+                    returnError = error.errorFor(field)
+                    if (returnError != null) {
+                        return returnError
+                    }
+                }
+            }
+        }
+        return null
+    }
+
+    override fun toString(): String {
+        return "BraintreeError for $field: $message -> ${fieldErrors?.toString() ?: ""}"
+    }
+
+    internal constructor()
+
+    override fun describeContents(): Int {
+        return 0
+    }
+
+    override fun writeToParcel(dest: Parcel, flags: Int) {
+        dest.writeString(field)
+        dest.writeString(message)
+        dest.writeTypedList(fieldErrors)
+    }
+
+    protected constructor(`in`: Parcel) {
```

Comment:
Nitpick - can we organize the constructors to be next to each other? Also, moving the constructors to be below the variable definitions at the top of the file.

Reply:
On second thought these make the most sense in order because the parcel order in `writeToParcel` and the `constructor(in: Parcel)` order need to match exactly.

In Kotlin we can use [Parcelize](https://developer.android.com/kotlin/parcelize) in the future to remove some of the boilerplate required for `Parcelable`. We may be able to do it in the current major version, but as a precaution it's probably better to delay that change until the next major version when we make everything a `data` object.